Improve API Request Validations

betsyecastro opened the pull request:
Implements the following API request validation rules:

- Person (slug): Must contain only alphanumeric characters and periods.
- Search fields: Must be at least three characters long.
- School: Must match one of the whitelisted school short names.
- With data: Allowed only when the person parameter is present.

betsyecastro committed:
Update API request validations

betsyecastro committed:
Adds a separate validation rule for allowed schools

betsyecastro committed:
Refactors with_data validation into a dedicated method

betsyecastro committed:
Adds feature tests for invalid ProfilesApiRequest parameters

betsyecastro committed:
Implements ValidSearchString rule for search fields

betsyecastro committed:
Rejects API requests containing unexpected parameters

shukla-m reviewed (approved):
Looks good! I tested the updates locally and it seemed to work as expected. See [comment](https://github.com/utdal/profiles/pull/194#issuecomment-2956990740) regarding `person` query parameter with invalid value, in case it causes any issues.

betsyecastro committed:
Updates API Request to allow schools in different formats and to use built in validation rules for search fields

betsyecastro committed:
Removes ValidSearchString file

wunc committed:
Merge branch 'develop' into improve-api-request-validations

wunc committed:
API validation tweaks

- use more thorough search pattern regex (for example, some profile slugs may have dashes and quotes)
- add `search_section` and `accepting_undergrad` as valid params
- simplify the `with_data` validation by using the built-in `Rule::prohibitedIf()`
- also allow `with_data` if `data_type` is filled
- for extra param checking, use the `after()` method instead of `passedValidation()`.
